[c10d] Fix async error in batch_isend_irecv

aashaka opened the pull request:
Summary:
`batch_isend_irecv` previously required the use of `torch.cuda.synchronize` to avoid data race conditions. This was because the ncclStreams were recorderd in the returned ncclWork object _before_ a ncclGroupEnd by the `_batch_p2p_manager` was issued. Thus, the `req.wait()` was effectively waiting on nothing, leading to the later operators working on incorrect intermediate data.

This fix:
- keeps track of ncclStreams to wait on, and records them in the work objects after the batch manager issues a ncclGroupEnd
- renames the `_batch_p2p_manager` to `_coalescing_manager` for generality
- removes the explicit check for NCCL backend inside `_batch_p2p_manager` in distributed_c10.py and moves the manager start/end to ProcessGroup.hpp, in order to transparently work with all process groups

Test Plan: Modified the unittest for `batch_isend_irecv` to check that received tensors are the same as expected tensors. Verified that the test fails before the change, and passes after the change.

Differential Revision: D38100789

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1405,6 +1405,34 @@ ProcessGroupNCCL::Options::Options(bool is_high_priority_stream)
+
+  int batch_idx = 0;
+  for (const auto& req : reqs) {
+    auto ncclWork = dynamic_cast<ProcessGroupNCCL::WorkNCCL*>(req.get());
```
can we static_cast<> here so we have compile-time guarantees and don't blow up at runtime?

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -272,9 +272,14 @@ class TORCH_API ProcessGroupNCCL : public ProcessGroup {
 
   const std::string getBackendName() const override {
-      return std::string(NCCL_BACKEND_NAME);
+        return std::string(NCCL_BACKEND_NAME);
```
nit: extra space here

kwen2501 reviewed (approved):
LGTM.
My previous comments in [D38100789](https://www.internalfb.com/diff/D38100789) have been addressed.
Just added a minor format comment in this PR.

Commit:
[c10d] Fix async error in batch_isend_irecv (#82450)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82450

`batch_isend_irecv` previously required the use of `torch.cuda.synchronize` to avoid data race conditions. This was because the ncclStreams were recorderd in the returned ncclWork object _before_ a ncclGroupEnd by the `_batch_p2p_manager` was issued. Thus, the `req.wait()` was effectively waiting on nothing, leading to the later operators working on incorrect intermediate data.

This fix:
- keeps track of ncclStreams to wait on, and records them in the work objects after the batch manager issues a ncclGroupEnd
- renames the `_batch_p2p_manager` to `_coalescing_manager` for generality
- removes the explicit check for NCCL backend inside `_batch_p2p_manager` in distributed_c10.py and moves the manager start/end to ProcessGroup.hpp, in order to transparently work with all process groups

Test Plan: Modified the unittest for `batch_isend_irecv` to check that received tensors are the same as expected tensors. Verified that the test fails before the change, and passes after the change.

Reviewed By: kwen2501

Differential Revision: D38100789

fbshipit-source-id: 200a414c560368616c2ecd82e453f01647ba93c1

kwen2501 commented:
@pytorchbot merge
